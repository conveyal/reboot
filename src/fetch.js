import fetch from 'isomorphic-fetch'
import isObject from 'lodash.isobject'
import {createAction} from 'redux-actions'

export const INCREMENT_FETCH = 'increment outstanding fetches'
export const DECREMENT_FETCH = 'decrement outstanding fetches'
export const FETCH = 'fetch'
export const FETCH_MULTIPLE = 'fetch multiple'
export const FETCH_ERROR = 'fetch error'

export const incrementFetches = createAction(INCREMENT_FETCH)
export const decrementFetches = createAction(DECREMENT_FETCH)
export const fetchAction = createAction(FETCH)
export const fetchMultiple = createAction(FETCH_MULTIPLE)
export const fetchError = createAction(FETCH_ERROR)

export function middleware (store) {
  return (next) => (action) => {
    if (action.type === FETCH) {
      return store.dispatch(runFetchAction(action.payload, store.getState()))
    } else if (action.type === FETCH_MULTIPLE) {
      return store.dispatch(runFetchMultiple(action.payload, store.getState()))
    } else {
      return next(action)
    }
  }
}

export default fetchAction

/**
 * Calls fetch, adds Auth and Content header if needed. Automatically parses content based on type.
 *
 * @returns Promise
 */

function runFetch ({
  options = {},
  retry = false,
  url
}, state) {
  const isJSON = isObject(options.body)
  return fetch(url, {
    ...options,
    body: isJSON ? JSON.stringify(options.body) : options.body,
    headers: {
      ...createAuthorizationHeader(state),
      ...createContentHeader(isJSON),
      ...(options.headers || {})
    }
  })
    .then(checkStatus)
    .then(createResponse)
    .then(async (response) =>
      (retry && await retry(response))
        ? runFetch({options, retry, url}, state)
        : response)
}

function runFetchAction ({
  next,
  options = {},
  retry = false,
  url
}, state) {
  return [
    incrementFetches(),
    runFetch({options, retry, url}, state)
      .then((response) => [decrementFetches(), next(null, response)])
      .catch((error) =>
        createErrorResponse(error)
          .then((response) => [decrementFetches(), fetchError(response), next(error, response)]))
  ]
}

/**
 * @returns Promise
 */

function runFetchMultiple ({
  fetches,
  next
}, state) {
  return [
    incrementFetches(),
    Promise.all(fetches.map((fetch) => runFetch(fetch, state)))
      .then((responses) => [decrementFetches(), next(null, responses)])
      .catch((error) =>
        createErrorResponse(error)
          .then((response) => [decrementFetches(), fetchError(response), next(error, response)]))
  ]
}

function createAuthorizationHeader (state) {
  return state.user && state.user.idToken
    ? {Authorization: `bearer ${state.user.idToken}`}
    : {}
}

function checkStatus (res) {
  if (res.status >= 200 && res.status < 300) {
    return res
  } else {
    throw res
  }
}

function createContentHeader (isJSON) {
  return isJSON
    ? {'Accept': 'application/json', 'Content-Type': 'application/json;charset=UTF-8'}
    : {}
}

function createErrorResponse (res) {
  return res.headers
    ? createResponse(res)
    : Promise.resolve(res)
}

function createResponse (res) {
  return deserialize(res)
    .then((value) => ({
      url: res.url,
      status: res.status,
      statusText: res.statusText,
      headers: res.headers,
      value
    }))
    .catch((err) => ({
      value: err
    }))
}

function deserialize (res) {
  const header = [res.headers.get('Content-Type'), res.headers.get('Content')].filter(Boolean)
  for (let i = 0; i < header.length; i++) {
    if (header[i].indexOf('application/json') > -1) return res.json()
    if (header[i].indexOf('application/ld+json') > -1) return res.json()
    if (header[i].indexOf('application/octet-stream') > -1) return res.arrayBuffer()
  }
  return res.text()
}
