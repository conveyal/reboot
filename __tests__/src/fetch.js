// @flow
import nock from 'nock'

import fetch, {
  abortAllFetches,
  abortFetch,
  ABORT_FETCH_FAILED,
  ABORTED_FETCH,
  fetchMultiple,
  getID
} from '../../src/fetch'
import createStore from '../../src/store'

const URL = 'http://fakeurl.com'

describe('fetch', () => {
  it('should dispatch a fetch action and call next', (done) => {
    const store = createStore()

    nock(URL)
      .get('/')
      .reply(200, {})

    const action = fetch({
      url: URL,
      next: (err, res) => done(err, res)
    })

    store.dispatch(action)
  })

  it('should automatically serialize json bodies', (done) => {
    const store = createStore()

    const body = {
      data: 'important'
    }

    nock(URL)
      .post('/', body)
      .reply(200, (path, requestBody) => {
        expect(requestBody === JSON.stringify(body)).toBeTruthy()
        return 'good'
      })

    const action = fetch({
      url: URL,
      options: {
        method: 'post',
        body
      },
      next (response) {
        response.text().then(value => {
          expect(value).toEqual('good')
          done()
        })
      }
    })

    store.dispatch(action)
  })

  it('should work with FormData', (done) => {
    const store = createStore()
    const data = new window.FormData()
    data.append('name', 'world')

    nock(URL)
      .post('/')
      .reply(200, 'FormData')

    const action = fetch({
      url: URL,
      options: {
        method: 'post',
        body: data
      },
      next: (error, response) => {
        response.text().then(value => {
          expect(value).toEqual('FormData')
          done(error)
        })
      }
    })

    store.dispatch(action)
  })

  it('should automatically parse json responses', (done) => {
    const store = createStore()
    nock(URL)
      .get('/')
      .reply(200, {hello: 'world'})

    const action = fetch({
      url: URL,
      next: (error, response) => {
        expect(response.value).toEqual({hello: 'world'})
        done(error)
      }
    })

    store.dispatch(action)
  })

  it('should fetch multiple urls in one go', (done) => {
    const store = createStore()
    nock(URL)
      .get('/one')
      .reply(200, 'one', {'Content-Type': 'text/plain'})

    nock(URL)
      .get('/two')
      .reply(200, 'two', {'Content-Type': 'text/plain'})

    const action = fetchMultiple({
      fetches: [{
        url: `${URL}/one`
      }, {
        url: `${URL}/two`
      }],
      next: (error, responses) => {
        expect(responses.length).toBe(2)
        expect(responses[0].value).toBe('one')
        expect(responses[1].value).toBe('two')
        done(error)
      }
    })

    store.dispatch(action)
  })

  it('should retry based on the response', (done) => {
    const store = createStore()
    let triedOnce = false
    nock(URL)
      .get('/')
      .reply((uri, body) => {
        if (triedOnce) {
          return [200]
        } else {
          triedOnce = true
          return [202]
        }
      })

    const action = fetch({
      url: URL,
      retry: (response) => response.status !== 200,
      next: (error, response) => {
        expect(triedOnce).toBeTruthy()
        done(error)
      }
    })

    store.dispatch(action)
  })

  it('should allow retrying with an async function', (done) => {
    const store = createStore()
    let triedOnce = false
    nock(URL)
      .get('/')
      .reply((uri, body) => {
        if (triedOnce) {
          return [200]
        } else {
          triedOnce = true
          return [202]
        }
      })

    const action = fetch({
      url: URL,
      retry: (response) => {
        return new Promise((resolve, reject) => {
          if (response.status !== 200) {
            setTimeout(() => resolve(true), 10)
          } else {
            resolve(false)
          }
        })
      },
      next: (error, response) => {
        expect(triedOnce).toBeTruthy()
        done(error)
      }
    })

    store.dispatch(action)
  })

  it('should dispatch a fetchError if the arity of `next` is < 2', async () => {
    const store = createStore()
    nock(URL)
      .get('/')
      .reply(400, 'ERROR')

    const action = fetch({
      url: URL,
      next (response) {
        throw new Error('Should not be executed')
      }
    })

    const actionResult = store.dispatch(action)
    await actionResult[1]
    expect(store.getActions()).toHaveLength(3) // inc, dec, and fetch error
  })

  it('should not dispatch a fetchError if the arity of `next` is >= 2', (done) => {
    const store = createStore()
    nock(URL).get('/').reply(400, 'ERROR')

    const action = fetch({
      url: URL,
      next (error, response) {
        if (!error) {
          done('Expected error')
        }
      }
    })

    const actionResult = store.dispatch(action)
    Promise.resolve(actionResult[1])

    // let the actions finish
    setTimeout(() => {
      expect(store.getActions()).toMatchSnapshot() // no fetch error action
      done()
    }, 100)
  })

  describe('abort', () => {
    it('should not call next on a aborted fetch', (done) => {
      const type = 'TEST'
      const id = getID()
      const store = createStore()

      nock(URL)
        .get('/')
        .delay(1)
        .reply(200, 'Done')

      const action = fetch({
        type,
        id,
        url: URL,
        next: () => done('Should not be called')
      })

      store.dispatch(action)
      store.dispatch(abortFetch({type, id}))

      setTimeout(() => {
        expect(store.getActions()).toContainEqual({
          type: ABORTED_FETCH,
          payload: {type, id}
        })
        done()
      }, 2)
    })

    it('should call next if fetch finishes before abort', (done) => {
      const type = 'TEST'
      const id = getID()
      const store = createStore()

      nock(URL)
        .get('/')
        .reply(200, 'done')

      const next = jest.fn()
      store.dispatch(fetch({
        id,
        type,
        url: URL,
        next: () => {
          next()
        }
      }))

      setTimeout(() => {
        store.dispatch(abortFetch({type, id}))
        expect(store.getActions()).toContainEqual({
          type: ABORT_FETCH_FAILED,
          payload: {type, id}
        })
        expect(next).toHaveBeenCalled()
        done()
      }, 2)
    })

    it('should cancel all active fetches of any type', (done) => {
      const store = createStore()

      nock(URL)
        .get('/')
        .delay(1)
        .reply(200, 'done')

      const fetchNum = 10
      for (let i = 0; i < fetchNum; i++) {
        store.dispatch(fetch({
          type: `test-${i}`,
          url: URL,
          next: () => done('should not be called')
        }))
      }

      store.dispatch(abortAllFetches())
      expect(store.getActions()).toHaveLength(fetchNum * 3)
      done()
    })

    it('should abort active fetches of the same type', (done) => {
      const store = createStore()
      const type = 'TEST'
      const fetchNum = 10
      const ids = []

      for (let i = 0; i < fetchNum; i++) {
        nock(URL)
          .get(`/item/${i}`)
          .delay(1)
          .reply(200, 'done')

        const id = getID()
        ids.push(id)

        store.dispatch(fetch({
          type,
          id,
          url: `${URL}/item/${i}`,
          next: () => {
            // should be called once
            done()
          }
        }))
      }

      for (let i = 0; i < (fetchNum - 1); i++) {
        expect(store.getActions()).toContainEqual({
          type: ABORTED_FETCH,
          payload: {type, id: ids[i]}
        })
      }
    })
  })
})
