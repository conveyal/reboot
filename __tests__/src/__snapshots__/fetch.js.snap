// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetch should dispatch a fetchError if the arity of \`next\` is < 2 1`] = `
Array [
  Object {
    "payload": Object {
      "options": Object {},
      "url": "http://fakeurl.com",
    },
    "type": "increment outstanding fetches",
  },
  Object {
    "payload": Object {
      "options": Object {},
      "url": "http://fakeurl.com",
    },
    "type": "decrement outstanding fetches",
  },
  Object {
    "payload": Body {
      "_abort": false,
      "_raw": Array [],
      "body": PassThrough {
        "_events": Object {
          "close": [Function],
          "drain": [Function],
          "error": [Function],
          "finish": [Function],
          "prefinish": [Function],
          "unpipe": [Function],
        },
        "_eventsCount": 6,
        "_maxListeners": undefined,
        "_readableState": ReadableState {
          "awaitDrain": 0,
          "buffer": BufferList {
            "head": null,
            "length": 0,
            "tail": null,
          },
          "decoder": null,
          "defaultEncoding": "utf8",
          "destroyed": false,
          "emittedReadable": false,
          "encoding": null,
          "endEmitted": false,
          "ended": false,
          "flowing": null,
          "highWaterMark": 16384,
          "length": 0,
          "needReadable": true,
          "objectMode": false,
          "pipes": null,
          "pipesCount": 0,
          "readableListening": false,
          "reading": false,
          "readingMore": false,
          "resumeScheduled": false,
          "sync": false,
        },
        "_transformState": Object {
          "afterTransform": [Function],
          "needTransform": false,
          "transforming": false,
          "writecb": null,
          "writechunk": null,
          "writeencoding": null,
        },
        "_writableState": WritableState {
          "bufferProcessing": false,
          "bufferedRequest": null,
          "bufferedRequestCount": 0,
          "corked": 0,
          "corkedRequestsFree": Object {
            "entry": null,
            "finish": [Function],
            "next": null,
          },
          "decodeStrings": true,
          "defaultEncoding": "utf8",
          "destroyed": false,
          "ended": false,
          "ending": false,
          "errorEmitted": false,
          "finalCalled": false,
          "finished": false,
          "highWaterMark": 16384,
          "lastBufferedRequest": null,
          "length": 0,
          "needDrain": false,
          "objectMode": false,
          "onwrite": [Function],
          "pendingcb": 0,
          "prefinished": false,
          "sync": true,
          "writecb": null,
          "writelen": 0,
          "writing": false,
        },
        "allowHalfOpen": true,
        "readable": true,
        "writable": true,
      },
      "bodyUsed": false,
      "headers": Headers {
        "_headers": Object {},
      },
      "ok": false,
      "size": 0,
      "status": 400,
      "statusText": "Bad Request",
      "timeout": 0,
      "url": "http://fakeurl.com",
      "value": undefined,
    },
    "type": "fetch error",
  },
]
`;

exports[`fetch should not dispatch a fetchError if the arity of \`next\` is >= 2 1`] = `
Array [
  Object {
    "payload": Object {
      "options": Object {},
      "url": "http://fakeurl.com",
    },
    "type": "increment outstanding fetches",
  },
  Object {
    "payload": Object {
      "options": Object {},
      "url": "http://fakeurl.com",
    },
    "type": "decrement outstanding fetches",
  },
]
`;
